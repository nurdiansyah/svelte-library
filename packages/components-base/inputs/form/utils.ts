// Supports determination of isControlled().
// Controlled input accepts its current value as a prop.
//
// @param value
// @returns {boolean} true if string (including '') or number (including zero)
export function hasValue(value) {
  return value != null && !(Array.isArray(value) && value.length === 0);
}

// Determine if field is empty or filled.
// Response determines if label is presented above field or as placeholder.
//
// @param value
// @param SSR
// @returns {boolean} False when not present or empty string.
//                    True when any number or string with length.
export function isFilled(value: any, SSR = false) {
  return hasValue(value) && value !== "";
}
