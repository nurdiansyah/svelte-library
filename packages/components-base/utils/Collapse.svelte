<script lang="ts">
  import { clsx } from "./clsx";

  import { createEventDispatcher } from "svelte";
  import { slide } from "svelte/transition";
  const noop = () => undefined;

  export let show: boolean = false;
  let { class: className } = $$props;
  let classes: string | undefined;
  export let onEntering: (e?: Event) => any = noop;
  export let onEntered: (e?: Event) => any = noop;
  export let onExiting: (e?: Event) => any = noop;
  export let onExited: (e?: Event) => any = noop;

  $: classes = clsx(className, "dbx-collapse", show && "-show");
</script>

<div
  {...$$restProps}
  class={classes}
  transition:slide|local
  on:introstart
  on:introend
  on:outrostart
  on:outroend
  on:introstart={onEntering}
  on:introend={onEntered}
  on:outrostart={onExiting}
  on:outroend={onExited}>
  <slot />
</div>

<style lang="scss" global>
  .dbx-collapse:not(.-show) {
    display: none;
  }
</style>
