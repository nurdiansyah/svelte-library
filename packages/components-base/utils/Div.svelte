<script>
  export let node = undefined;
</script>

<div {...$$props} bind:this={node}>
  <slot />
</div>
