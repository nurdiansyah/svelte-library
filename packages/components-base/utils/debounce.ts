// Corresponds to 10 frames at 60 Hz.
// A few bytes payload overhead when lodash/debounce is ~3 kB and debounce ~300 B.
export function debounce(func: (arg0: any, arg1: any[]) => void, wait = 166) {
  let timeout: NodeJS.Timeout;
  function debounced(...args: any[]) {
    // eslint-disable-next-line consistent-this
    const that = this;
    const later = () => {
      func.apply(that, args);
    };
    clearTimeout(timeout);
    timeout = setTimeout(later, wait);
  }

  debounced.clear = () => {
    clearTimeout(timeout);
  };

  return debounced;
}
