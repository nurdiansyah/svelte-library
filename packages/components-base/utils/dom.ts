export const ownerDocument = (node: Node) => (node && node.ownerDocument) || document;
export const ownerWindow = (node: Node) => {
  const doc = ownerDocument(node);
  return doc.defaultView || window;
};

// A change of the browser zoom change the scrollbar size.
// Credit https://github.com/twbs/bootstrap/blob/3ffe3a5d82f6f561b82ff78d82b32a7d14aed558/js/src/modal.js#L512-L519
export const getScrollbarSize = () => {
  const scrollDiv = document.createElement("div");
  scrollDiv.style.width = "99px";
  scrollDiv.style.height = "99px";
  scrollDiv.style.position = "absolute";
  scrollDiv.style.top = "-9999px";
  scrollDiv.style.overflow = "scroll";

  document.body.appendChild(scrollDiv);
  const scrollbarSize = scrollDiv.offsetWidth - scrollDiv.clientWidth;
  document.body.removeChild(scrollDiv);

  return scrollbarSize;
};

export const getOriginalBodyPadding = () => {
  const style = window ? window.getComputedStyle(document.body, null) : { getPropertyValue: (e?: any) => undefined };

  return parseInt((style && style.getPropertyValue("padding-right")) || "0", 10);
};

export const setScrollbarWidth = (padding) => {
  if (padding > 0) {
    document.body.style.paddingRight = `${padding}px`;
  }
};

export const isBodyOverflowing = () => (window ? document.body.clientWidth < window.innerWidth : false);
