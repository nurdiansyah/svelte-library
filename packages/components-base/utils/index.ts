export { default as Collapse } from "./Collapse.svelte";
export { default as Scrollbar } from "./Scrollbar.svelte";
export * as actions from "./actions";
export * from "./generate";
export * from "./style";
export * from "./popper";
