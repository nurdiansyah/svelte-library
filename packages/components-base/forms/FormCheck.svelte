<script lang="ts">
  import clsx from "clsx";
  import { getId } from "../utils";

  let className: string | null = null,
    classes: string;
  export { className as class };
  export let disabled: boolean = false;
  export let inline: boolean = false;
  export let type: "radio" | "checkbox" | "switch" = "checkbox";
  export let id = getId();

  $: classes = clsx(className, "form-check", inline && "-inline", disabled && "-disabled");
</script>

{#if type === 'radio'}
  <input {...$$restProps} id="rad-{id}" type="radio" {disabled} />
{:else}
  <input {...$$restProps} id="cb-{id}" {type} {disabled} />
{/if}

<style lang="scss" global>
  @use "./form-check";
</style>
