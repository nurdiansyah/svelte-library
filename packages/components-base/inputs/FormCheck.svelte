<script lang="ts">
  import type { FormStore } from "./form/FormStore";
  import type { ValueCheck } from "../types/input";

  import { onMount } from "svelte";
  import { clsx } from "../utils";
  import { setContext } from "./form/form-control-context";
  import { getContext } from "./form/form-context";
  import Radio from "./Radio.svelte";
  import CheckBox from "./Checkbox.svelte";
  import Switch from "./Switch.svelte";
  import FormLabel from "./FormLabel.svelte";
  import Checkbox from "./Checkbox.svelte";

  /**
   * value data component FormCheck
   * @typedef {Object} ValueCheck
   * @property {string} label - label data
   * @property {*} value - value data
   * @property {boolean=} disabled - non aktifkan data
   * @property {string=} hint
   */
  // props
  /**
   * @type {('checkbox'|'radio'|'switch')}
   */
  export let type: "checkbox" | "radio" | "switch" = "checkbox";

  /**
   * @type {string}
   */
  export let name: string;

  /**
   * @type {boolean} [disabled=false]
   */
  export let disabled: boolean = false;

  /**
   * @type {ValueCheck[]} [selected[]]
   */
  export let selected: ValueCheck[] = [];

  /**
   *
   * @type {ValueCheck[]} [values=[]]
   */
  export let values: ValueCheck[] = [];

  /**
   *
   * @type {string[]} [errors[]]
   */
  export let errors: string[] = [];

  /**
   * @type {function(value: Value): void} [validate] - validate function
   */
  export let validate: ((value: any) => void) | undefined = undefined;

  /**
   *
   * @type {'small'} [size]
   */
  export let size: "small" | undefined = undefined;

  /**
   *
   * @type {boolean}
   */
  export let touched: boolean = false;

  // state
  let context = getContext();
  let { class: className, "data-testid": dataTestId } = $$props;
  let classes;
  let fieldState;
  let formStore: FormStore;
  let Component: any;
  delete $$restProps["data-testid"]; // remove variables node data-testid

  // handler
  onMount(() => {
    formStore = context.formStore;
    formStore.createField({ name, validate, value: selected, errors, touched });
    fieldState = formStore.getField(name);
  });

  /**
   *
   * @param {*} value
   */
  function isChecked(value: any) {
    if (selected.includes(value)) {
      return true;
    }
    return false;
  }

  function handleChange(e) {}

  // reactive
  $: classes = clsx(className, "form-check");
  $: {
    switch (type) {
      case "switch":
        Component = Switch;
        break;
      case "radio":
        Component = Radio;
        break;
      default:
        Component = Checkbox;
        break;
    }
  }
</script>

<div class={classes} data-testid={dataTestId}>
  {#each values as valueCheck}
    <FormLabel
      label={valueCheck.label}
      disabled={valueCheck.disabled}
      error={errors && errors.length > 0}
      let:idFor={_id}>
      <svelte:component
        this={Component}
        id={_id}
        {...$$restProps}
        disabled={disabled || valueCheck.disabled}
        checked={isChecked(valueCheck.value)}
        {size} />
    </FormLabel>
  {/each}
</div>
