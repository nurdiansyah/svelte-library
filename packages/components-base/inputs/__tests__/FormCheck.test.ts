import { itConformance } from "../../utils/test-utils";
import FormCheck from "./features/MockFormCheck.svelte";

describe("FormCheck", () => {
  const testId = "form-check";
  const props = { "data-testid": testId };

  itConformance(FormCheck, { props });
});
