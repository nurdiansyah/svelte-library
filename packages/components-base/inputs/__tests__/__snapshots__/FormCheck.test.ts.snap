// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheck cek render komponen tanpa error dan snapshot 1`] = `
<body>
  <div>
    <div
      class="form-check"
      data-testid="form-check"
    >
      <label
        class="form-label -color-primary"
        for="dbx-ZF_Qg"
      >
        <span
          class="input-check"
        >
          <button
            class="-icon button-base"
          >
            <span
              class="label"
            >
              <input
                class="input"
                id="dbx-ZF_Qg"
                type="checkbox"
              />
               
              <svg
                class="dbx-icon -type-svg"
                role="img"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
                <title>
                  CheckBoxOutlineBlank
                </title>
              </svg>
               
              
              
            </span>
          </button>
           
          <div
            class="ripple-container"
          />
          
           
        </span>
         
         
        <span
          class="label"
        >
          foo
        </span>
      </label>
      <label
        class="form-label -color-primary"
        for="dbx-rt_L7"
      >
        <span
          class="input-check"
        >
          <button
            class="-icon button-base"
          >
            <span
              class="label"
            >
              <input
                class="input"
                id="dbx-rt_L7"
                type="checkbox"
              />
               
              <svg
                class="dbx-icon -type-svg"
                role="img"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
                <title>
                  CheckBoxOutlineBlank
                </title>
              </svg>
               
              
              
            </span>
          </button>
           
          <div
            class="ripple-container"
          />
          
           
        </span>
         
         
        <span
          class="label"
        >
          foo2
        </span>
      </label>
      <label
        class="form-label -color-primary"
        for="dbx-c_-78"
      >
        <span
          class="input-check"
        >
          <button
            class="-icon button-base"
          >
            <span
              class="label"
            >
              <input
                class="input"
                id="dbx-c_-78"
                type="checkbox"
              />
               
              <svg
                class="dbx-icon -type-svg"
                role="img"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
                <title>
                  CheckBoxOutlineBlank
                </title>
              </svg>
               
              
              
            </span>
          </button>
           
          <div
            class="ripple-container"
          />
          
           
        </span>
         
         
        <span
          class="label"
        >
          foo3
        </span>
      </label>
      <label
        class="form-label -color-primary"
        for="dbx-7F4Ls"
      >
        <span
          class="input-check"
        >
          <button
            class="-icon button-base"
          >
            <span
              class="label"
            >
              <input
                class="input"
                id="dbx-7F4Ls"
                type="checkbox"
              />
               
              <svg
                class="dbx-icon -type-svg"
                role="img"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                />
                <title>
                  CheckBoxOutlineBlank
                </title>
              </svg>
               
              
              
            </span>
          </button>
           
          <div
            class="ripple-container"
          />
          
           
        </span>
         
         
        <span
          class="label"
        >
          foo4
        </span>
      </label>
    </div>
  </div>
</body>
`;
